File: src/main/java/org/taktik/icure/security/database/DatabaseUserDetails.java
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.security.database;

import org.springframework.security.core.GrantedAuthority;
import org.taktik.icure.security.AbstractUserDetails;
import org.taktik.icure.security.PermissionSetIdentifier;

import java.util.List;
import java.util.Map;
import java.util.Set;

public class DatabaseUserDetails extends AbstractUserDetails {
	private static final long serialVersionUID = 1L;

	private String rev;
	private String passwordHash;
	private Map<String,String> applicationTokens;
	private String application;
	private String secret;
	private boolean use2fa;

	public DatabaseUserDetails(PermissionSetIdentifier permissionSetIdentifier, Set<GrantedAuthority> authorities, String passwordHash, String secret, Boolean use2fa) {
		super(permissionSetIdentifier, authorities);
		this.passwordHash = passwordHash;
		this.secret = secret;
		this.use2fa = use2fa != null ? use2fa : false;
	}

	public String getRev() {
		return rev;
	}

	public void setRev(String rev) {
		this.rev = rev;
	}

	@Override
	public String getPassword() {
		return passwordHash;
	}

	public Map<String, String> getApplicationTokens() {
		return applicationTokens;
	}

	public void setApplicationTokens(Map<String, String> applicationTokens) {
		this.applicationTokens = applicationTokens;
	}

	public void setApplication(String application) {
		this.application = application;
	}

	public String getApplication() {
		return application;
	}

	public String getSecret() {
		return secret;
	}

	public boolean isUse2fa() {
		return use2fa;
	}

}

File: src/main/kotlin/org/taktik/icure/security/database/DatabaseUserDetails.kt
/*
 * Copyright (c) 2020. Taktik SA, All rights reserved.
 */
package org.taktik.icure.security.database

import org.springframework.security.core.GrantedAuthority
import org.taktik.icure.entities.security.Permission
import org.taktik.icure.security.PermissionSetIdentifier
import org.taktik.icure.security.AbstractUserDetails

class DatabaseUserDetails(
        permissionSetIdentifier: PermissionSetIdentifier,
        authorities: Set<GrantedAuthority>,
        principalPermissions: Set<Permission>,
        val passwordHash: String?,
        val secret: String?,
        val use2fa: Boolean = false,
        val rev: String? = null,
        val applicationTokens: Map<String, String> = mapOf(),
        val application: String? = null,
        val groupIdUserIdMatching: List<String> = listOf()
) : AbstractUserDetails(
    permissionSetIdentifier, authorities, principalPermissions
) {
    override fun getPassword(): String? {
        return passwordHash
    }

    @Suppress("DuplicatedCode")
    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is DatabaseUserDetails) return false
        if (!super.equals(o)) return false

        if (passwordHash != o.passwordHash) return false
        if (secret != o.secret) return false
        if (use2fa != o.use2fa) return false
        if (rev != o.rev) return false
        if (applicationTokens != o.applicationTokens) return false
        if (application != o.application) return false
        if (groupIdUserIdMatching != o.groupIdUserIdMatching) return false

        return true
    }

    @Suppress("DuplicatedCode")
    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + passwordHash.hashCode()
        result = 31 * result + secret.hashCode()
        result = 31 * result + use2fa.hashCode()
        result = 31 * result + (rev?.hashCode() ?: 0)
        result = 31 * result + (applicationTokens?.hashCode() ?: 0)
        result = 31 * result + (application?.hashCode() ?: 0)
        result = 31 * result + (groupIdUserIdMatching?.hashCode() ?: 0)
        return result
    }

}
